### PR TITLE
feat: support type cast in SchemaAdapter

### LINKED ISSUE #6381: Support Type Coercion in SchemaAdapter
### Is your feature request related to a problem or challenge?

Currently SchemaAdapter::map_projections returns an error if the provided `file_schema` contains columns of a different type to the expected `table_schema`.

In order to accomodate schema evolution, it would be advantageous to be able to instead cast the array data from the file schema to the table schema where possible.

### Describe the solution you'd like

```
impl SchemaAdapter {

    fn map_schema(&self, file_schema: &Schema) -> Result<SchemaMapping> {
        ...
    }
}

struct SchemaMapping {
    ...
}

impl SchemaMapping {
    fn map_batch(batch: RecordBatch) -> Result<RecordBatch> {
        ...
    }
}
```

Where `map_schema` would use [can_cast_types](https://docs.rs/arrow-cast/latest/arrow_cast/cast/fn.can_cast_types.html) to determine if conversion is possible, and `map_batch` would use [cast](https://docs.rs/arrow-cast/latest/arrow_cast/cast/fn.cast.html) to perform this conversion

### Describe alternatives you've considered

We could not do this

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by e1ijah1**
# Which issue does this PR close?

Closes #6381

# Rationale for this change

# What changes are included in this PR?

introduces a `map_schema` method that maps the file schema to the table schema, and a `SchemaMapping` structure to hold the mapping and necessary type conversions. 

# Are these changes tested?

Yes

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 1]
```diff
@@ -450,6 +451,69 @@ impl SchemaAdapter {
...
+    fn map_batch(&self, batch: RecordBatch) -> Result<RecordBatch> {
+        let mut mapped_cols = Vec::with_capacity(self.field_mappings.len());
+
+        for (idx, data_type) in &self.field_mappings {
+            let array = batch.column(*idx);
+            let casted_array = arrow::compute::cast(array, data_type)?;
+            mapped_cols.push(casted_array);
+        }
+
+        let record_batch = RecordBatch::try_new(self.table_schema.clone(), mapped_cols)?;
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/mod.rs:514]: I think this should use try_new_with_options to match adapt_batch above

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 2]
```diff
@@ -1124,6 +1191,138 @@ mod tests {
...
+
+        assert_eq!(mapped_batch.schema(), table_schema);
+        assert_eq!(mapped_batch.num_columns(), 3);
+        assert_eq!(mapped_batch.num_rows(), 2);
+
+        let c1 = mapped_batch
+            .column(0)
+            .as_any()
+            .downcast_ref::<StringArray>()
+            .unwrap();
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/mod.rs:1306]:
```suggestion
            .as_string::<i32>()
```

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 3]
```diff
@@ -1124,6 +1191,138 @@ mod tests {
...
+        let c2 = mapped_batch
+            .column(1)
+            .as_any()
+            .downcast_ref::<UInt32Array>()
+            .unwrap();
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/mod.rs:1311]:
```suggestion
            .as_primitive::<UInt32Type>();
```

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 4]
```diff
@@ -1124,6 +1191,138 @@ mod tests {
...
+        let c3 = mapped_batch
+            .column(2)
+            .as_any()
+            .downcast_ref::<Float64Array>()
+            .unwrap();
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/mod.rs:1316]:
```suggestion
            .as_primitive::<Float64Type>();
```

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 5]
```diff
@@ -1124,6 +1191,138 @@ mod tests {
...
+
+        // file schema lacks necessary columns
+        let file_schema = Schema::new(vec![
+            Field::new("c1", DataType::Utf8, true),
+            Field::new("c2", DataType::Int32, true),
+        ]);
+
+        let mapping = adapter.map_schema(&file_schema);
+
+        assert!(
```
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/mod.rs:1250]: We could perhaps do `unwrap_err().to_string()` and check the error message (to ensure that it is failing for the right reasons

**Review by tustvold — APPROVED** [review]
This is really nice, thank you. Left some minor comments, but this is also good to go as is

**Comment by e1ijah1** [comment]
> This is really nice, thank you. Left some minor comments, but this is also good to go as is

Thank you for taking the time to review my PR 🙏. I have updated the code. Please take another look.

**Review by tustvold — APPROVED** [review]
Thank you

**Pull request merged**